### DISCUSSION TITLE
More generic python ws server/client

### DISCUSSION THREAD

**Discussion by nickroci** [perspective-dev/perspective#1571] [Ideas]
Since tornado is kind of compatible with asyncio would it be useful to make a non tornado specific client and server? I whipped something up quickly that seems to work OK with starlette for instance

**Comment by sc1f** [comment]
That would be great - I haven't heard of Starlette until now but it looks really interesting. I think the direction here for the generic websocket server is probably akin to what we do in the generic [async client](https://github.com/finos/perspective/blob/master/python/perspective/perspective/client/client.py) - a base class that handles a lot of the intricacies (especially that around handling binary transfer/chunked transfer), and a `send(msg)` interface that the framework-specific handlers need to override.

What you've built out for Starlette would be a great start - we appreciate any and all PRs/contributions!

**Reply by nickroci** [reply]
I do think eventually having a transport/framework agnostic layer could be good, with ability to plug in client and server. There is something to be said for an easy way to stream updating apache arrow tables between things and then visualise what is going on at any point. Especially as there seems to be interesting stuff going on with datafusion.

**Comment by nickroci** [comment]
I made a start to show it can be done, not ready for prime time but shows the concept I hope :) 

https://github.com/nickpholden/perspective/tree/starlette-ws
